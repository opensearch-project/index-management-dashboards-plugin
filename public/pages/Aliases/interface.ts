/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
export interface IAlias {
  alias: string;
  index: string;
  filter: string;
  "routing.index": string;
  "routing.search": string;
  is_write_index: string;
  indexArray: string[];
}
