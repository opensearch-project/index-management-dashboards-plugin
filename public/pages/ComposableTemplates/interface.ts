/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { IComposableTemplateRemote } from "../../../models/interfaces";

export interface ICatComposableTemplate {
  name: string;
  component_template: IComposableTemplateRemote;
  usedBy: string[];
  associatedCount: number;
}
