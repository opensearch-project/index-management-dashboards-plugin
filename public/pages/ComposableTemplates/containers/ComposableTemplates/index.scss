.ISM-component-templates-table .euiTableRow:hover {
    .icon-hover-info {
        color: $ouiColorPrimary
    }
    .icon-hover-danger {
        color: $ouiColorDanger
    }
}