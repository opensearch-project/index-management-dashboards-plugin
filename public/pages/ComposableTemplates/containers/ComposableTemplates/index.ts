/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ComposableTemplates from "./ComposableTemplates";

export default ComposableTemplates;
