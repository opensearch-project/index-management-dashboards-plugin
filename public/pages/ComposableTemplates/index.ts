/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ComposableTemplates from "./containers/ComposableTemplates";

export default ComposableTemplates;
