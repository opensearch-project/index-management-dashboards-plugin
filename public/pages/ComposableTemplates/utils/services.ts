/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonService } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";

export const getAllUsedComponents = async ({ commonService }: { commonService: CommonService }) => {
  const allTemplatesResponse = await commonService.apiCaller<{
    index_templates?: Array<{
      name: string;
      index_template: TemplateItemRemote;
    }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: "/_index_template/*",
    },
  });

  if (allTemplatesResponse && allTemplatesResponse.ok) {
    return allTemplatesResponse.response.index_templates || [];
  }

  return [];
};
