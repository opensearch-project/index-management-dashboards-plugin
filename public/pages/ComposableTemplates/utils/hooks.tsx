/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { CoreStart } from "opensearch-dashboards/public";
import { useState, useCallback, useContext, useEffect, useRef } from "react";
import { CommonService, ServicesContext } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";
import { getAllUsedComponents } from "./services";
import { BrowserServices } from "../../../models/interfaces";

export const getTemplate = async (props: {
  templateName: string;
  commonService: CommonService;
  coreService: CoreStart;
}): Promise<TemplateItemRemote> => {
  const response = await props.commonService.apiCaller<{
    index_templates: Array<{ name: string; index_template: TemplateItemRemote }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: `/_index_template/${props.templateName}?flat_settings=true`,
    },
  });
  let error: string = "";
  if (response.ok) {
    const findItem = response.response?.index_templates?.find((item) => item.name === props.templateName);
    if (findItem) {
      const templateDetail = findItem.index_template;
      return JSON.parse(JSON.stringify(templateDetail));
    }
    error = `The template [${props.templateName}] does not exist.`;
  } else {
    error = response.error || "";
  }

  props.coreService.notifications.toasts.addDanger(error);
  throw new Error(error);
};

export const getComponentMapTemplate = async ({ commonService }: { commonService: CommonService }) => {
  const allIndicesTemplates = await getAllUsedComponents({
    commonService,
  });

  const componentMapTemplate: Record<string, string[]> = {};

  allIndicesTemplates.forEach((item) => {
    item.index_template.composed_of?.forEach((composedItem) => {
      componentMapTemplate[composedItem] = componentMapTemplate[composedItem] || [];
      componentMapTemplate[composedItem].push(item.name);
    });
  });

  return componentMapTemplate;
};

const cache: {
  usedRequest: Promise<Record<string, string[]>> | undefined;
} = {
  usedRequest: undefined,
};

let listenerNumber = 0;

const CACHE_REFRESH_EVENT = "CACHE_REFRESH_EVENT";

export const getComponentMapTemplateWithCache = (
  props: Parameters<typeof getComponentMapTemplate>[0] & {
    force?: boolean;
  }
) => {
  if (props.force || !cache.usedRequest) {
    cache.usedRequest = getComponentMapTemplate({
      commonService: props.commonService,
    });
  }
  return cache.usedRequest;
};

export const submitTemplateChange = async (props: {
  templateName: string;
  commonService: CommonService;
  coreService: CoreStart;
  transformTemplate: (templateItem: TemplateItemRemote) => TemplateItemRemote;
}) => {
  const { templateName, commonService, coreService, transformTemplate = (templateItem) => templateItem } = props;
  const currentTemplate = await getTemplate({
    templateName,
    commonService,
    coreService,
  });
  const updateResult = await commonService.apiCaller({
    endpoint: "transport.request",
    data: {
      method: "POST",
      path: `/_index_template/${templateName}`,
      body: transformTemplate(currentTemplate),
    },
  });
  return updateResult;
};

export const useComponentMapTemplate = () => {
  const [componentMapTemplate, setComponentMapTemplate] = useState<Record<string, string[]>>({});
  const [loading, setLoading] = useState(false);
  const services = useContext(ServicesContext) as BrowserServices;
  const destroyRef = useRef<boolean>(false);

  const reload = useCallback(
    (force?: boolean) => {
      setLoading(true);
      getComponentMapTemplateWithCache({
        force,
        commonService: services.commonService,
      }).then((res) => {
        if (force) {
          window.dispatchEvent(new CustomEvent(CACHE_REFRESH_EVENT));
        }
        if (destroyRef.current) {
          return;
        }
        setComponentMapTemplate(res);
        setLoading(false);
      });
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const eventHandler = useCallback(() => {
    reload();
  }, [reload]);

  useEffect(
    () => {
      reload();
      window.addEventListener(CACHE_REFRESH_EVENT, eventHandler);
      listenerNumber++;
      return () => {
        window.removeEventListener(CACHE_REFRESH_EVENT, eventHandler);
        listenerNumber--;
        if (listenerNumber === 0) {
          /**
           * when all the listeners are gone
           * we should clear the usedRequest
           * so that the componentTemplates page
           * will re-fetch the index templates data
           */
          cache.usedRequest = undefined;
        }
      };
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [eventHandler]
  );

  useEffect(() => {
    return () => {
      destroyRef.current = true;
    };
  }, []);

  return {
    reload,
    loading,
    componentMapTemplate,
  };
};
