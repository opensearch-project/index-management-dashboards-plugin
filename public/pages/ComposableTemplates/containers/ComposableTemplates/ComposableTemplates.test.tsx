/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { act, render, waitFor } from "@testing-library/react";
import { Redirect, Route, Switch } from "react-router-dom";
import { HashRouter as Router } from "react-router-dom";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import ComposableTemplates from "./index";
import { ServicesContext } from "../../../../services";
import { ROUTES } from "../../../../utils/constants";
import { CoreServicesContext } from "../../../../components/core_services";
import userEvent from "@testing-library/user-event";
import { ICatComposableTemplate } from "../../interface";
import { getApplication, getNavigationUI, getUISettings } from "../../../../services/Services";

jest.mock("../../../../services/Services", () => ({
  ...jest.requireActual("../../../../services/Services"),
  getUISettings: jest.fn(),
  getApplication: jest.fn(),
  getNavigationUI: jest.fn(),
}));

beforeEach(() => {
  (getUISettings as jest.Mock).mockReturnValue({
    get: jest.fn().mockReturnValue(false), // or false, depending on your test case
  });
  (getApplication as jest.Mock).mockReturnValue({});

  (getNavigationUI as jest.Mock).mockReturnValue({});
});

function renderWithRouter() {
  return {
    ...render(
      <Router>
        <Switch>
          <Route
            path={ROUTES.COMPOSABLE_TEMPLATES}
            render={(props) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <ServicesContext.Provider value={browserServicesMock}>
                  <ComposableTemplates {...props} />
                </ServicesContext.Provider>
              </CoreServicesContext.Provider>
            )}
          />
          <Route path={ROUTES.CREATE_COMPOSABLE_TEMPLATE} render={() => <>create template</>} />
          <Redirect from="/" to={ROUTES.COMPOSABLE_TEMPLATES} />
        </Switch>
      </Router>
    ),
  };
}

const testTemplateId = "test";

describe("<ComposableTemplates /> spec", () => {
  beforeEach(() => {
    browserServicesMock.commonService.apiCaller = jest.fn(async (payload) => {
      if (payload.endpoint === "transport.request" && (payload?.data?.method === "GET" || "")) {
        if (payload?.data?.path.includes("_component_template") && payload?.data?.path.includes(testTemplateId)) {
          return {
            ok: false,
            error: "Not Found1",
          };
        } else if (payload.data?.path?.startsWith("/_index_template")) {
          return {
            ok: true,
            response: {
              index_templates: [
                {
                  name: "test_template",
                  index_template: {
                    composed_of: [`${testTemplateId}-1`],
                  },
                },
              ],
            },
          };
        }

        return {
          ok: true,
          response: {
            component_templates: [
              {
                name: `${testTemplateId}-2`,
                component_template: {
                  template: {
                    aliases: {
                      a: {},
                    },
                  },
                },
              },
              {
                name: testTemplateId,
                component_template: {
                  template: {
                    aliases: {
                      a: {},
                    },
                  },
                },
              },
              {
                name: `${testTemplateId}-1`,
                component_template: {
                  template: {},
                },
              },
            ] as ICatComposableTemplate[],
          },
        };
      }

      return {
        ok: true,
        response: {},
      };
    }) as any;
    window.location.hash = "/";
  });
  it("renders the component", async () => {
    const { container, getByTestId } = renderWithRouter();

    expect(container.firstChild).toMatchSnapshot();
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(2);
    });
    userEvent.click(getByTestId("tableHeaderCell_name_0").querySelector("button") as Element);
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(3);
    });
  });

  it("with some actions", async () => {
    const { getByPlaceholderText, getByTestId, findByTestId, findByText, queryByTestId } = renderWithRouter();

    expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(2);
    await userEvent.click(getByTestId("tableHeaderCell_name_0").querySelector('[data-test-subj="tableHeaderSortButton"]') as Element);
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(3);
    });
    /**
     * search
     */
    await userEvent.type(getByPlaceholderText("Search..."), `${testTemplateId}{enter}`);
    await waitFor(() => {
      expect(queryByTestId(`templateDetail-${testTemplateId}`)).toBeNull();
    });
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(5);
      expect(coreServicesMock.notifications.toasts.addDanger).toBeCalledTimes(1);
    });
    await userEvent.clear(getByPlaceholderText("Search..."));
    await findByTestId(`templateDetail-${testTemplateId}`);
    await userEvent.click(document.querySelector(".euiFilterButton-hasNotification") as Element);
    await userEvent.click(getByTestId(`FilterGroupSelectItem-aliases`));
    await waitFor(() => {
      expect(queryByTestId(`checkboxSelectRow-${testTemplateId}-1`)).toBeNull();
    });
    await act(async () => {
      await userEvent.click(document.body);
    });
    await waitFor(() => {
      expect(queryByTestId(`FilterGroupSelectItem-aliases`)).toBeNull();
    });

    /**
     * sort
     */
    await userEvent.click(getByTestId(`checkboxSelectRow-${testTemplateId}`));
    await waitFor(() => {
      expect(getByTestId("deleteAction")).toBeEnabled();
    });

    /**
     * view associated
     */
    await userEvent.click(getByTestId(`ViewAssociatedIndexTemplates-${testTemplateId}`));
    await findByTestId("euiFlyoutCloseButton");
    await userEvent.click(getByTestId("euiFlyoutCloseButton"));

    /**
     * delete
     */
    await userEvent.click(getByTestId(`DeleteComponentTemplate-${testTemplateId}`));
    await findByTestId("deleteConfirmButton");
    await act(async () => {
      await userEvent.click(getByTestId("deleteConfirmButton"));
    });
    await waitFor(() => {
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledWith(`Delete [${testTemplateId}] successfully`);
    });

    /**
     * create
     */
    await userEvent.click(getByTestId("Create component templateButton"));
    await findByText("create template");
  });

  it("render without component templates", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn(async (payload) => {
      if (payload.endpoint === "transport.request") {
        return {
          ok: true,
          response: {
            component_templates: [] as ICatComposableTemplate[],
          },
        };
      }

      return {
        ok: true,
        response: {},
      };
    }) as any;
    const { getByPlaceholderText, findByText, getByText, getByTestId } = renderWithRouter();
    await findByText("You have no templates.");
    await userEvent.type(getByPlaceholderText("Search..."), `${testTemplateId}{enter}`);
    await findByText("There are no templates matching your applied filters. Reset your filters to view your templates.");
    await userEvent.click(getByText("Reset filters"));
    await findByText("You have no templates.");
    await userEvent.click(getByTestId("CreateComponentTemplateWhenNoTemplateFound"));
    await findByText("create template");
  });
});
