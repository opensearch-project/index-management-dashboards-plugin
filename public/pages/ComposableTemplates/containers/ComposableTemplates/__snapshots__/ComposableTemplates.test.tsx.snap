// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComposableTemplates /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 10px;"
  >
    <div
      class="euiFlexItem"
    >
      <span
        class="euiTitle euiTitle--medium"
      >
        Component templates
      </span>
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            id="some_html_id-help-0"
          >
            <div>
              Component templates are reusable building blocks for composing index or data stream templates. You can define component templates with common index configurations and associate them to an index template.
               
              <button
                class="euiLink euiLink--primary"
                type="button"
              >
                Learn more
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <button
                class="euiButton euiButton--danger euiButton-isDisabled"
                data-test-subj="deleteAction"
                disabled=""
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Delete
                  </span>
                </span>
              </button>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <button
                class="euiButton euiButton--primary euiButton--fill"
                data-test-subj="Create component templateButton"
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Create component template
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 5px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFormControlLayout euiFormControlLayout--fullWidth"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldSearch euiFieldSearch--fullWidth"
              placeholder="Search..."
              type="search"
              value=""
            />
            <div
              class="euiFormControlLayoutIcons"
            >
              <span
                class="euiFormControlLayoutCustomIcon"
              >
                EuiIconMock
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
        style="flex-basis: 100px;"
      >
        <div
          class="euiFilterGroup"
        >
          <div
            class="euiPopover euiPopover--anchorDownCenter"
          >
            <div
              class="euiPopover__anchor"
            >
              <button
                class="euiButtonEmpty euiButtonEmpty--text euiFilterButton euiFilterButton-hasNotification euiFilterButton--hasIcon"
                type="button"
              >
                <span
                  class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                >
                  EuiIconMock
                  <span
                    class="euiButtonEmpty__text euiFilterButton__text-hasNotification"
                  >
                    <span
                      class="euiFilterButton__textShift"
                      data-text="Types"
                      title="Types"
                    >
                      Types
                    </span>
                    <span
                      aria-label="3 available filters"
                      class="euiNotificationBadge euiNotificationBadge--medium euiNotificationBadge--subdued euiFilterButton__notification"
                    >
                      3
                    </span>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      class="euiBasicTable euiBasicTable-loading ISM-component-templates-table"
      data-test-subj="templatesTable"
    >
      <div>
        <div
          class="euiTableHeaderMobile"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiCheckbox"
              >
                <input
                  aria-label="Select all rows"
                  class="euiCheckbox__input"
                  disabled=""
                  id="_selection_column-checkbox_some_html_id"
                  type="checkbox"
                />
                <div
                  class="euiCheckbox__square"
                />
                <label
                  class="euiCheckbox__label"
                  for="_selection_column-checkbox_some_html_id"
                >
                  Select all rows
                </label>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiTableSortMobile"
              >
                <div
                  class="euiPopover euiPopover--anchorDownRight"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        EuiIconMock
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Sorting
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <table
          class="euiTable euiTable--responsive"
          id="some_html_id"
          tabindex="-1"
        >
          <caption
            class="euiScreenReaderOnly euiTableCaption"
          />
          <thead>
            <tr>
              <th
                class="euiTableHeaderCellCheckbox"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <div
                    class="euiCheckbox euiCheckbox--inList euiCheckbox--noLabel"
                  >
                    <input
                      aria-label="Select all rows"
                      class="euiCheckbox__input"
                      data-test-subj="checkboxSelectAll"
                      disabled=""
                      id="_selection_column-checkbox_some_html_id"
                      type="checkbox"
                    />
                    <div
                      class="euiCheckbox__square"
                    />
                  </div>
                </div>
              </th>
              <th
                aria-live="polite"
                aria-sort="descending"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_name_0"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Name"
                    >
                      Name
                    </span>
                    EuiIconMock
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_templateTypes_1"
                role="columnheader"
                scope="col"
              >
                <span
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Type"
                  >
                    Type
                  </span>
                </span>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_component_template._meta.description_2"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Description"
                    >
                      Description
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_associatedCount_3"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Associated index templates"
                    >
                      Associated index templates
                    </span>
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                role="columnheader"
                scope="col"
              >
                <span
                  class="euiTableCellContent euiTableCellContent--alignRight"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Actions"
                  >
                    Actions
                  </span>
                </span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="euiTableRow"
            >
              <td
                class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                colspan="6"
              >
                <div
                  class="euiTableCellContent euiTableCellContent--alignCenter"
                >
                  <span
                    class="euiTableCellContent__text"
                  >
                    No items found
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
